api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java

```diff
@@ -153,4 +178,84 @@ public void test_nonUniqueIds() {
         new Institution().shortName("test").displayName("We are all individuals"));
     service.createInstitution(new Institution().shortName("test").displayName("I'm not"));
   }
+
+  @Test
+  public void test_emailValidation_domain() {
+    final Institution inst =
+        service.createInstitution(
+            new Institution()
+                .shortName("Broad")
+                .displayName("The Broad Institute")
+                .emailDomains(Lists.newArrayList("broad.org", "lab.broad.org")));
+
+    final DbUser user = createUser("user@broad.org");
+    final DbVerifiedInstitutionalAffiliation affiliation =
+        createAffiliation(user, inst.getShortName());
+
+    assertThat(service.validate(affiliation, user.getContactEmail())).isTrue();
+  }
+
+  @Test
+  public void test_emailValidation_address() {
+    final Institution inst =
+        service.createInstitution(
+            new Institution()
+                .shortName("Broad")
+                .displayName("The Broad Institute")
+                .emailDomains(Lists.newArrayList("broad.org", "mit.edu"))
+                .emailAddresses(
+                    Lists.newArrayList("external-researcher@sanger.uk", "science@aol.com")));
+
+    final DbUser user = createUser("external-researcher@sanger.uk");
+    final DbVerifiedInstitutionalAffiliation affiliation =
+        createAffiliation(user, inst.getShortName());
+
+    assertThat(service.validate(affiliation, user.getContactEmail())).isTrue();
+  }
+
+  @Test
+  public void test_emailValidation_null() {
+    final Institution inst =
+        service.createInstitution(
+            new Institution().shortName("Broad").displayName("The Broad Institute"));
+
+    final DbUser user = userDao.save(new DbUser());
```

Comment:
This might be made more realistic / meaningful by having a user with a valid email address which doesn't exist in the institution's list. None of the tests explicitly checks this important negative case yet, while a totally-empty dbUser object is less-likely to exist in our system.

Reply:
agreed